### PR TITLE
Randomized Data Iterator

### PR DESCRIPTION
**Pull request by neeleshdodda44**
Previously, the training and test sets were formed by iterating each datapoint of each dataset in forwards and backwards order, respectively. 

Now, before the iterators are created, a mapping is created such that the index of each datapoint is randomly mapped to either the training set or the test set. Then, the iterator only adds datapoints that belong to its corresponding set, using the mapping. This way, there is now __no bias__ in how datapoints are assigned to the training/test sets.

Major changes:
- `create_mapping` is new method of `iterators.py` that maps data points to training/test set given a datapoint count
- All of the create iterator methods have been modified to use this mapping as an argument
- The iterator tests and `Runner` class have been modified to create a mapping and pass it in when creating iterators.

### CODE REVIEW

**Comment by kiddyboots216** [comment]
1) There's no need to reinvent the wheel when there's a somewhat accepted format for creating a schedule; see https://github.com/berkeleydeeprlcourse/homework/blob/master/hw3/dqn_utils.py ctrl+f 'schedule' (this is still a constant time solution, I believe, but avoids computation at datapoint selection time which is relevant given that our amoritzed case is not too different from significant precomputation -nonetheless this space complexity should be avoided when possible as the point of a lot of the code we're writing is to conserve RAM)
2) The motivation for supporting a random iterator is to be able to use Stochastic Gradient Descent. To this extent, I don't understand why supporting our own random dataset iteration is necessary when our https://github.com/georgymh/decentralized-ml/blob/master/custom/keras.py currently supports all Keras optimizers, including SGD.

**Comment by neeleshdodda44** [comment]
Summarizing changes:

- Relaxed memory constraints. New functionality for random data iterator runs in `O(n)` space complexity, where `n` is the number of datapoints.
- `labeler` is back to being a string referring to the column name of the column that is the label
- At start of each epoch, loaded dataset is shuffled.
- Used `scikit-learn` and `numpy` to generate the train test split and batches instead of iterating over each datapoint.
- Separate tests for randomized data iterator
- Updated Runner to use randomized data iterator
- Updated tests for Runner to use string `labeler`
- Updated DMLJob to use string `labeler`

@georgymh feel free to review from here.

**Comment by kiddyboots216** [comment]
no scikit-learn please unless there is an INCREDIBLE reason for using it. you literally cannot install scikit-learn on windows machines half the time. sklearn.train_test_split can be recreated given pandas and numpy.

**Comment by neeleshdodda44** [comment]
Makes sense, I'll go ahead implement train-test without `scikit-learn` (honestly just a few more lines of code).

File: tests/test_dataset_manager.py [hunk 1]
```diff
@@ -35,19 +35,23 @@ def same_raw_data_test(dsm, expected_test1_raw, expected_test2_raw):
...
-        expected_test2_transformed.equals(actual_test1_transformed)
+    actual_test_transformed = transform_dsm['train'].append(transform_dsm['test'])
+    actual_test_transformed = actual_test_transformed.reset_index(drop=True)
+    merged_data = pd.merge(
+        actual_test_transformed,
+        expected_test_transformed,
+        on = ['a', 'b', 'c'],
+        how = 'inner'
+    )
+    assert len(expected_test_transformed) == len(actual_test_transformed)
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply tests/test_dataset_manager.py]: So the point of these assertions are to:
1. Check that the total number of data points has not changed.
2. Check that the data points themselves have not changed. 

Admittedly, I should `assert` that the length of the training set is `split*length` and `(1-split)*length` for the test set (will add). However, since the order of the data points in the training/tests sets has changed, it is inefficient to check if each data point in one is found in the other (since originally the dataframes were guaranteed to have data points in the same order, `df.equals` could be used). By checking that the length of the inner merge is still the same, we ensure that the dataframes have the same data points (but in different order).

Tl;dr it may seem relaxed, but the test itself has not changed.

File: tests/test_dataset_manager.py [hunk 2]
```diff
@@ -35,19 +35,23 @@ def same_raw_data_test(dsm, expected_test1_raw, expected_test2_raw):
     assert expected_test1_raw.equals(actual_test1_raw)
     assert expected_test2_raw.equals(actual_test2_raw)
 
-def accurate_transform_test(dsm, expected_test1_transformed, expected_test2_transformed):
+def accurate_transform_test(dsm, expected_test_transformed):
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply tests/test_dataset_manager.py]: You're right, see my comment above. I will assert to make sure the split is correct.

File: tests/artifacts/configuration.ini [hunk 3]
```diff
@@ -1,5 +1,5 @@
 [GENERAL]
-dataset_path = datasets/mnist
+dataset_path = tests/artifacts/runner_scheduler
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply tests/artifacts/configuration.ini]: Well you're right, this is only going to be used by the runner/scheduler, and I should put this in the corresponding folder.

But I wanted to point out that there shouldn't be a global `configuration.ini`, since pretty much all tests should use their own, or none at all.

File: core/runner.py [hunk 4]
```diff
@@ -39,6 +40,8 @@ def __init__(self, config_manager):
         config = config_manager.get_config()
         self.iden = str(uuid.uuid4())[:8]
         self.dataset_path = config.get("GENERAL", "dataset_path")
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/runner.py]: Yes (I think we talked about this in person later).

File: core/dataset_manager.py [hunk 5]
```diff
@@ -165,7 +178,7 @@ def clean_up(self):
         """
         if self.tfp:
             shutil.rmtree(self.tfp)
-        assert not os.path.isdir(self.tfp)
+            assert not os.path.isdir(self.tfp)
```
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/dataset_manager.py]: Subtle, but if the test failed, then an error would be thrown because `self.tfp` was `None` -_-

File: core/runner.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (neeleshdodda44) [reply core/runner.py]: This will be done as a later task though, as we discussed.
